### PR TITLE
Bump cluster-api jobs for v1.26

### PR DESCRIPTION
**Pull request by aniruddha2000**
part of: https://github.com/kubernetes-sigs/cluster-api/issues/7671

Signed-off-by: Aniruddha Basak <codewithaniruddha@gmail.com>

### CODE REVIEW

**Comment by sbueringer** [comment]
/hold

I have some findings, but I need some time for a proper review (properly first week in January)

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-main-upgrades.yaml
Comment (fabriziopandini) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-main-upgrades.yaml:285]: We have to keep 1-24-1-25 and add a new job for 1-25-1-26
(same for all the branches, main, 1.2, and also 1.3)

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-main.yaml
Comment (fabriziopandini) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-main.yaml:257]: KUBERNETES_VERSION_UPGRADE_FROM and KUBERNETES_VERSION_UPGRADE_TO must be changed, eventually also ETCD_VERSION_UPGRADE_TO and COREDNS_VERSION_UPGRADE_TO
(same for all the branches, not for main and 1.2 only)

**Review by fabriziopandini — COMMENTED** [review]
Left some comments.

Also, periodic upgrade-1-25-latest must be changed into upgrade-1-26-latest
(same for all the branches, main, 1.2, and also 1.3)

Taking a look at the prior art https://github.com/kubernetes/test-infra/pull/27421 might help (note that 1.3 did not exist at the time)

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-main-upgrades.yaml
Comment (aniruddha2000) [reply config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-main-upgrades.yaml:285]: Got it!

So, now we have `1-24-1-25`, `1-25-1-26`. What about the `1-25-latest`, we need to change it `1-26-latest` right?

Comment (fabriziopandini) [reply config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-main-upgrades.yaml:332]: The end state must be: `1-24-1-25`, `1-25-1-26`, `1-26-latest` (that is equivale to testing upgrades to all the stable releases + the upgrade to latest)

**Comment by fabriziopandini** [comment]
/lgtm
@sbueringer for a final pass

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-main.yaml
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-main.yaml:286]:
```suggestion
            value: "ci/latest-1.27"
```

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-release-1-2.yaml
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-release-1-2.yaml:282]:
```suggestion
            value: "ci/latest-1.27"
```

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-release-1-3.yaml
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-release-1-3.yaml:276]:
```suggestion
            value: "ci/latest-1.27"
```

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2-upgrades.yaml
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2-upgrades.yaml:334]: @fabriziopandini  Q: This PR looks like we're planning to support 1.26 with release-1.2. I'm not sure if that's the plan (xref: https://github.com/kubernetes-sigs/cluster-api/issues/7671#issuecomment-1362728064)

(It even goes one step further by also testing v1.26 => v1.27 with CAPI v1.2.x)

Comment (sbueringer) [reply config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2-upgrades.yaml:334]: @aniruddha2000 I talked to Fabrizio and we would support v1.26 in v1.2 & v1.3. So you can ignore this comment.

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-3-upgrades.yaml
Comment (ykakarap) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-3-upgrades.yaml:352]: The image used here should be the same as the other tests: `gcr.io/k8s-staging-test-infra/kubekins-e2e:v20221221-461b6105cf-1.25`

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2-upgrades.yaml
Comment (ykakarap) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2-upgrades.yaml:307]: The image used here should be the same as other tests: `gcr.io/k8s-staging-test-infra/kubekins-e2e:v20221221-461b6105cf-1.24`.

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2.yaml
Comment (ykakarap) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2.yaml:137]: This will start a CAPI v1.1.5 cluster with Kubernetes version v1.26 which is not part of our support matrix. I realise that we did this for the previous kubernetes versions but just curious if we should continue doing this?

**Review by ykakarap — CHANGES_REQUESTED** [review]
LGTM pending addressing the comments.

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2.yaml
Comment (sbueringer) [reply config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2.yaml:137]: I think it's okay, given that it keeps the job in test-infra in sync with the code inside of the branch (we will bump this in the release-1.2 branch when we cherry-pick the 1.26 change into release-1.2).

This will be the last time we have to do this in test-infra as the additional upgrade jobs have been moved into clusterctl_upgrade_test.go

Comment (sbueringer) [reply config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2.yaml:137]: Just to clarify. It will create a v1.26 cluster with Cluster API v1.2.x and then it will deploy Cluster API v1.1.5 on top of that v1.26 cluster. It won't create a v1.26 cluster with Cluster API v1.1.x

(still outside of the support matrix to run Cluster API v1.1.x on top of v1.26, but just wanted to make sure we're talking about the same thing)

Comment (aniruddha2000) [reply config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2.yaml:137]: So I should be reverting back to 1.25?

Comment (sbueringer) [reply config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2.yaml:137]: Thought about it some more. Yuvaraj is right. Let's use 1.24 here. It's the latest version supported by 1.1

Comment (aniruddha2000) [reply config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2.yaml]: right

**Comment by aniruddha2000** [comment]
Waiting for a final review if anything still missing from my side 🙂

**Comment by fabriziopandini** [comment]
/lgtm
from my side, let's start getting a signal after the CAPI patch release + 1 or 2 days delay

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-3-upgrades.yaml
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-3-upgrades.yaml:352]:
```suggestion
      - image: gcr.io/k8s-staging-test-infra/kubekins-e2e:v20221221-461b6105cf-1.25
```

Please change the image and then rebase the entire PR to latest master + ensure that the new jobs are then using the same images as the other jobs in the same file. Thx!

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-3-upgrades.yaml
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-3-upgrades.yaml:352]:
```suggestion
      - image: gcr.io/k8s-staging-test-infra/kubekins-e2e:v20230105-c1beb6cf98-1.25
```
(to use the same image as the other periodics in this file)

File: config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2-upgrades.yaml
Comment (sbueringer) [inline config/jobs/kubernetes-sigs/cluster-api/cluster-api-periodics-release-1-2-upgrades.yaml:354]:
```suggestion
      - image: gcr.io/k8s-staging-test-infra/kubekins-e2e:v20230105-c1beb6cf98-1.24
```

**Comment by sbueringer** [comment]
To give some context:

Every Cluster API version is using a specific Kubernetes minor version (regarding go dependencies like client-go and also we use the same Go version as the Kubernetes version). The mapping is:
* main => v1.26
* v1.3 => v1.25
* v1.2 => v1.24

To get the right Go version (and maybe other stuff not sure) we use the `kubekins-e2e` images from upstream. That's why different branches are using different images.

The most relevant parts of the image name (`gcr.io/k8s-staging-test-infra/kubekins-e2e:v20230105-c1beb6cf98-1.24`) are:
* v20230105-c1beb6cf98 => "version of the image"
* 1.24 => for which Kubernetes version the image was build

In test-infra there is a job continuously bumping the "version of the image" via PRs (e.g. https://github.com/kubernetes/test-infra/pull/22831)

What we have to do is essentially:
* ensure we're using the image with the right Kubernetes version for each job
* ensure we're using the latest "version of the image" when we open PRs to not rollback changes of the bump job

**Comment by sbueringer** [comment]
@aniruddha2000 Thx for the patience. Should be the last findings :)

**Comment by aniruddha2000** [comment]
@sbueringer thanks for all the reviews! Now I know what is happening behind it and next time it will be easier for me and I will make fewer mistakes in the image hash part.

**Comment by sbueringer** [comment]
Thank you very much!!

/lgtm

/hold

Will merge after the cherry-pick PRsin CAPI. Probably tomorrow.

**Pull request merged**
